torch/testing/_internal/opinfo_helper.py

```diff
@@ -0,0 +1,120 @@
+import collections
+import warnings
+from functools import partial
+
+import torch
+from torch.testing._internal.common_cuda import (TEST_CUDA)
+import torch.testing._internal.common_methods_invocations as cmi
+from torch.testing._core import _dispatch_dtypes
+from torch.testing import (all_types_and_complex_and,
+                           all_types_and_complex,
+                           all_types_and_half,
+                           all_types,
+                           complex_types,
+                           floating_and_complex_types,
+                           floating_types_and_half,
+                           floating_types,
+                           integral_types,
+                           floating_types_and,
+                           floating_and_complex_types_and,
+                           integral_types_and,
+                           all_types_and,
+                           )
+
+COMPLETE_DTYPES_DISPATCH = (
+    all_types,
+    all_types_and_complex,
+    all_types_and_half,
+    floating_types,
+    floating_and_complex_types,
+    floating_types_and_half,
+    integral_types,
+    complex_types,
+)
+
+EXTENSIBLE_DTYPE_DISPATCH = (
+    all_types_and_complex_and,
+    floating_types_and,
+    floating_and_complex_types_and,
+    integral_types_and,
+    all_types_and,
+)
+
+# Better way to acquire devices?
+DEVICES = ['cpu'] + (['cuda'] if TEST_CUDA else [])
+
+# Class to tag the dynamically generated types.
+
+
+class _dynamic_dispatch_dtypes(_dispatch_dtypes):
+    pass
+
+
+def get_supported_dtypes(op, sample_inputs_fn, device):
+    # Returns the supported dtypes for the given operator and device pair.
+    assert device in ['cpu', 'cuda']
+    if not TEST_CUDA and device == 'cuda':
+        warnings.warn("WARNING: CUDA is not available, information pertaining to CUDA dtypes could be wrong")
```

Comment:
When 'cuda' is request but CUDA isn't available, should this just short-circuit and return an empty tuple of dtypes?

Reply:
Sorry, I meant: instead of throwing a warning? Or maybe it should error instead of short-circuit?

Reply:
I think it is better to return empty dtypes. 

If we `error` out here, this will conflict with always asking the user to use dynamic dtypes for `dtypes` and `dtypesIfCUDA`. 
https://github.com/pytorch/pytorch/pull/58698#discussion_r648059892

(If we make the change, acquiring dtypes for cuda will throw an error on CPU CI builds which will be terribly noisy!)

Reply:
That sounds good, but let's just inform the user that we're explicitly returning an empty dtypes entry.